### ISSUE TITLE
WeasyPrint 0.42 gets stuck

### ISSUE THREAD

**Issue by dhimmel** [Kozea/WeasyPrint#560]
We are calling WeasyPrint via pandoc. With 0.42 and a certain input, the command never completes, but does not throw an error, causing [builds to timeout](https://travis-ci.org/greenelab/scihub-manuscript/builds/329491074#L1864).

Downgrading from weasyprint 0.42 to 0.41 solves the issue: https://github.com/greenelab/scihub-manuscript/commit/5cb1245ca3e68c425b1364317e2496c17e20353c produced a [passing build](https://travis-ci.org/greenelab/scihub-manuscript/builds/329527648#L1421).

The issue doesn't happen for [all inputs](https://travis-ci.org/greenelab/deep-review/builds/329523064#L1373) (pandoc manuscripts). See for example, this passing build with WeasyPrint 0.42.

The command that fails [is](https://github.com/greenelab/scihub-manuscript/blob/a09d602ce3260746bd43476af4d2f58c1217d270/build/build.sh#L48-L62):

```sh
pandoc \
  --from=markdown \
  --to=html5 \
  --pdf-engine=weasyprint \
  --pdf-engine-opt=--presentational-hints \
  --filter=pandoc-fignos \
  --filter=pandoc-eqnos \
  --filter=pandoc-tablenos \
  --bibliography=$BIBLIOGRAPHY_PATH \
  --csl=$CSL_PATH \
  --metadata link-citations=true \
  --webtex=https://latex.codecogs.com/svg.latex? \
  --css=webpage/github-pandoc.css \
  --output=output/manuscript.pdf \
  $INPUT_PATH
```

Any ideas on what the problem could be or how to better diagnose the issue?

**Comment by liZe** [comment]
It may be a duplicate of #557. Could you please try to reproduce with master?

**Comment by dhimmel** [comment]
Will comment again when the [build](https://travis-ci.org/greenelab/scihub-manuscript/builds/329577050) for https://github.com/greenelab/scihub-manuscript/commit/cfe2a05dfa9355d601a438b450b3c5980e2299a1 is complete. So far not looking good (running for over 80 minutes), although its not clear yet exactly where its getting stuck. Interesting that it's not timing out after inactivity like before.

**Comment by dhimmel** [comment]
The build finally timed out after running for 8 hours. It appears the build got [stuck while running WeasyPrint](https://travis-ci.org/greenelab/scihub-manuscript/builds/329577050#L5412). So I don't think the issue has been fixed as of Kozea/WeasyPrint@ea9ffc9a3fcd6b45cc0172af4f9913abe1cb49e5.

**Comment by liZe** [comment]
Is it possible to get a HTML+CSS file generated by pandoc that makes the bug happen?

**Comment by dhimmel** [comment]
Currently I get the hangup locally when running the following with Python 3.6.4 and WeasyPrint 0.42 

```
weasyprint https://greenelab.github.io/scihub-manuscript/ weasyprint.pdf
```

Note that the content at https://greenelab.github.io/scihub-manuscript/ will change, so that URL may no longer trigger the issue in the future. If so, the versioned source for that webpage is [preserved here](https://github.com/greenelab/scihub-manuscript/tree/0f1a35706985507ab12ad7a8c3c97d99d6e4aaa0).

**Comment by liZe** [comment]
> I think we broke Travis hehe.

:smile:

> Currently I get the hangup locally when running the following with Python 3.6.4 and WeasyPrint 0.42
> weasyprint https://greenelab.github.io/scihub-manuscript/ weasyprint.pdf

I can't reproduce, even with the preserved version, probably because we don't have the same default font. Could you please tell me what launching `fc-match sans-serif` in a terminal gives on your system?

**Comment by dhimmel** [comment]
I'm on Ubuntu 17.10:

```
$ fc-match sans-serif
DejaVuSans.ttf: "DejaVu Sans" "Book"
```

**Comment by liZe** [comment]
> $ fc-match sans-serif
> DejaVuSans.ttf: "DejaVu Sans" "Book"

I've got the same default font, I don't know why it works for me. I've tried with both 0.42 and master, the PDF is correctly generated. I don't know how I could find what's going on…

**Comment by dhimmel** [comment]
I created a docker container, hoping that it would exhibit the error:

```sh
docker run \
  --name=weasyprint-560 \
  --interactive --tty \
  --entrypoint=bash \
  python:3.6
```

Then inside the container's bash shell, I ran:

```sh
pip install weasyprint==0.42
cd home
git clone --single-branch --branch gh-pages https://github.com/greenelab/scihub-manuscript.git
cd scihub-manuscript
git checkout 0f1a35706985507ab12ad7a8c3c97d99d6e4aaa0
weasyprint index.html weasyprint.pdf
```

This `weasyprint` command completed... i.e. no bug. This image is based on Debian 8. After lunch I will see if the Ubuntu image gets the error.

**Comment by dhimmel** [comment]
I switched to using conda to manage the environment in the Docker, so we can potentially better replicate our error:

```
docker run \
  --name=weasyprint-560 \
  --interactive --tty \
  --entrypoint=bash \
  continuumio/miniconda3:4.3.27
```

Then I ran:

```
apt-get install --yes gcc
cd /home
wget https://github.com/greenelab/manubot-rootstock/raw/59af0a2bdc23bbf48fae0acdcb8183888f12880e/build/environment.yml
conda env create --file=environment.yml
source activate manubot
git clone --single-branch --branch gh-pages https://github.com/greenelab/scihub-manuscript.git
cd scihub-manuscript
git checkout 0f1a35706985507ab12ad7a8c3c97d99d6e4aaa0
weasyprint index.html weasyprint.pdf
```

Unfortnately, `weasyprint` gives the following error:

```
OSError: dlopen() failed to load a library: cairo / cairo-2
```

Which seems to indicate that cairo is not installed (as per https://github.com/Kozea/CairoSVG/issues/84), although it seems that conda should be installing it. Anyways, I'll update if I make anymore progress.

**Comment by mengyyy** [comment]
After i update weasy[rint from 0.41 to 0.42 ,I found it could not generate pdf or png sfter it use 90% cpu for 30 minutes.
Python 3.5.2
Ubuntu 16.04

**Comment by liZe** [comment]
> After i update weasyprint from 0.41 to 0.42 ,I found it could not generate pdf or png sfter it use 90% cpu for 30 minutes.

@mengyyy Did you try the current `master` branch? You may have hit #557.

**Comment by charno6** [comment]
Hi, 

I am running into some issues with WeasyPrint getting stuck when creating PDFs from certain HTML files. I have now managed to (kind of) narrow down when this error occurs. My coding skills are insufficient to go into the WeasyPrint code to find out why this is happening; I hope this is helpful anyway. The files are attached.

[wp-table-demonstration.zip](https://github.com/Kozea/WeasyPrint/files/1692068/wp-table-demonstration.zip)

Basically, it's quite strange: On my system, the file "triangulate-error.html" will cause WeasyPrint to go into a kind of infinite loop when printing a page that contains a table; while "triangulate-noerror.html" will create a PDF within a matter of seconds. The only difference between them is in one additional "strong" tag in one of the cells. No error is logged by WeasyPrint. Interestingly, one other way to make the issue disappear is to change the font in the CSS to "Times New Roman". (I have also tried Georgia, Palatino, and my preferred font Iowan Old Style.)

My system is macOS Sierra 10.12.6 (16G1212), Python 3.6.4, WeasyPrint version 0.42.1.
I hope someone is able to reproduce this problem. If there is any way, I can help, let me know.

**Comment by liZe** [comment]
> I hope someone is able to reproduce this problem.

I can reproduce, thanks a lot for the example.

**Comment by dhimmel** [comment]
I updated the test PR (https://github.com/greenelab/scihub-manuscript/pull/39) to use the latest commit. This changed the build from timing out to failing. See the [error here](https://travis-ci.org/greenelab/scihub-manuscript/builds/337086117#L1906):

```
  File "/home/travis/miniconda/envs/manubot/lib/python3.6/site-packages/weasyprint/layout/inlines.py", line 201, in skip_first_whitespace
    result = skip_first_whitespace(box.children[index], next_skip_stack)
IndexError: list index out of range
Error producing PDF.
```

Haven't looked at this in detail but wanted to give a heads up.

**Comment by liZe** [comment]
> This changed the build from timing out to failing.

I'm really sorry :disappointed:. Your original URL raises the error, I'll fix that and add another regression test.

**Comment by liZe** [comment]
I've checked the tricky part of the new breaking line algorithm that causes these bugs (see #301 and #528). I've added some comments to help us in the future and corrected a couple of problems. I've also added a test to make sure this case won't happen again.

I really appreciate the time you take to report the issues and provide examples. If you find other crashes, please report them as well, I'll do my best to fix them as soon as possible!

**Comment by dhimmel** [comment]
> I really appreciate the time you take to report the issues and provide examples.

No worries! thanks for the fixes. I updated https://github.com/greenelab/scihub-manuscript/pull/39 to use WeasyPrint https://github.com/Kozea/WeasyPrint/commit/79e2b426a4a3a701d054a7973db19ce1ac956be7 and the build [succeeded](https://travis-ci.org/greenelab/scihub-manuscript/builds/337297833). So I think we're finally good!

**Comment by charno6** [comment]
I concur, thank you for the (rapid) fixes! I've just run WeasyPrint 0.42.2 against my list of articles that were previously causing problems and they all passed without a hitch. Thank you so much!

**Comment by liZe** [comment]
> Hi, I am running into the problem described above when using weasyprint 0.42.2.

Thank you for this report.

This issue has been closed with a commit fixing the original bug, so your problem is different even if it leads to the same consequences. Could you please open a separate issue?

> I would like to get this resolved as quickly as possible.

So do I!
